arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,738 @@
+
+  windowed_frame_averager( bool const should_round = false,
+                           unsigned const window_length = 20 );
+
+  // Process a new frame, returning the current frame average
+  void process_frame( vil_image_view< PixType > const& input,
+                      vil_image_view< PixType >& average ) override;
+
+  // Reset the internal average
+  void reset() override;
+
+  // Get number of frames used in the current window
+  size_t frame_count() const;
+
+private:
+  // Buffer containing pointers to last window_length frames
+  std::deque< vil_image_view< PixType > > window_buffer_;
+  size_t window_buffer_capacity_;
+};
+
+// ----------------------------------------------------------------------------
+// Shared functionality - process a frame while computing variance
+template < typename PixType >
+void
+online_frame_averager< PixType >
+::process_frame( vil_image_view< PixType > const& input,
+                 vil_image_view< PixType >& average,
+                 vil_image_view< double >& variance )
+{
+  // Check if this is the first time we have processed a frame of this size
+  if( dev1_tmp_space_.ni() != input.ni() ||
+      dev1_tmp_space_.nj() != input.nj() ||
+      dev1_tmp_space_.nplanes() != input.nplanes() )
+  {
+    dev1_tmp_space_.set_size( input.ni(), input.nj(), input.nplanes() );
+    dev2_tmp_space_.set_size( input.ni(), input.nj(), input.nplanes() );
+    variance.set_size( input.ni(), input.nj(), input.nplanes() );
+    variance.fill( 0.0 );
+    this->process_frame( input, average );
+    return;
+  }
+
+  // Calculate difference from last average
+  vil_math_image_abs_difference( input, this->last_average_, dev1_tmp_space_ );
+
+  // Update internal average
+  this->process_frame( input, average );
+
+  // Update the variance
+  vil_math_image_abs_difference( input, average, dev2_tmp_space_ );
+  vil_math_image_product( dev1_tmp_space_, dev2_tmp_space_, dev1_tmp_space_ );
+  variance.deep_copy( dev1_tmp_space_ );
+}
+
+// ----------------------------------------------------------------------------
+template < typename PixType >
+bool
+online_frame_averager< PixType >
+::has_resolution_changed( vil_image_view< PixType > const& input )
+{
+  return ( input.ni() != this->last_average_.ni() ||
+           input.nj() != this->last_average_.nj() ||
+           input.nplanes() != this->last_average_.nplanes() );
+}
+
+// ----------------------------------------------------------------------------
+// Helper function to allocate a completely new image, and cast the input image
+// to whatever specified type the output image is, scaling by some factor if
+// set and rounding if enabled, in one pass.
+template < typename inT, typename outT >
+void
+copy_cast( vil_image_view< inT > const& input, vil_image_view< outT >& output,
+           bool round )
+{
+  // Just deep copy if pixel formats equivalent and there is no scale factor
+  if( std::is_same< inT, outT >::value )
```

Comment:
...and if we were requiring C++17 (not yet, but maybe soon?), this could be `is_same_v< inT, outT >` instead of `is_same< inT, outT >::value`. :slightly_smiling_face: